Client/HttpCookieExtension.swift

```diff
@@ -0,0 +1,84 @@
+
+import Foundation
+import WebKit
+import Shared
+
+private let log = Logger.browserLogger
+
+public extension HTTPCookie {
+    
+    typealias Success = Bool
+    class func saveToDisk(_ filename: String = "CookiesData.json", completion: ((Success) -> Void)? = nil) {
+        let cookieStore = WKWebsiteDataStore.default().httpCookieStore
+        
+        /* For reason unkown the callback to getAllCookies is not called, when the save is done from Settings.
+         A possibility is https://bugs.webkit.org/show_bug.cgi?id=188242
+         Even with the issue being fixed it still sometimes doesn't work.
+         The network process is in suspened maybe?
+         
+         And for some reason fetch cookie records preemptively guarantees this works.
+         Best guess is that fetching records brings network process to active.
+         
+         Same applies to setting cookies back below.
+         */
+        WKWebsiteDataStore.default().fetchDataRecords(ofTypes: [WKWebsiteDataTypeCookies]) { (_) in}
+        cookieStore.getAllCookies { (cookies) in
+            let baseDir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
+            do {
+                let data = try NSKeyedArchiver.archivedData(withRootObject: cookies, requiringSecureCoding: false)
+                try data.write(to: URL(fileURLWithPath: "\(baseDir)/\(filename)"))
+            } catch {
+                log.error("Failed to write cookies to disk with error: \(error)")
+                completion?(false)
+                return
+            }
+            completion?(true)
+        }
+    }
+    
+    class func loadFromDisk(_ filename: String = "CookiesData.json", completion: ((Success) -> Void)? = nil) {
+        let baseDir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
+        do {
+            let data = try Data(contentsOf: URL(fileURLWithPath: "\(baseDir)/\(filename)"), options: Data.ReadingOptions.alwaysMapped)
+            if let cookies: [HTTPCookie] = try NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(data) as? [HTTPCookie] {
+                HTTPCookie.setCookies(cookies) { (success) in
+                    completion?(success)
+                }
+            } else {
+                log.error("Failed to load cookies from disk with error: Invalid data type")
+                completion?(false)
+            }
+        } catch {
+            log.error("Failed to load cookies from disk with error: \(error)")
+            completion?(false)
+        }
+    }
+    
+    private class func setCookies(_ cookies: [HTTPCookie], completion: ((Success) -> Void)?) {
+        let cookieStore = WKWebsiteDataStore.default().httpCookieStore
+        // For the purpose of the line below read the comment in saveCookies (in this same commit)
+        WKWebsiteDataStore.default().fetchDataRecords(ofTypes: [WKWebsiteDataTypeCookies]) { (_) in}
+        let dispatchGroup = DispatchGroup()
```

Comment:
What's your opinion on using `DispatchGroup` vs `Deferred` library here?
Your code is fine, I just wonder whether we should try to stick with one way to handle synchronization code like this

Reply:
Though `Deferred` is just a wrapper over dispatchGroups aka semaphores, in this particular we shouldn't use it because 
1. There can be a large number of cookies, and the extra function stack is not appreciable.
2. I want to keep it as native as possible.
3. Deferred does not provide a very out of the box solution for these cases. 
`private class func setCookies(_ cookies: [HTTPCookie]) -> Deferred<[Bool]> {
        let cookieStore = WKWebsiteDataStore.default().httpCookieStore
        // For the purpose of the line below read the comment in saveCookies (in this same commit)
        WKWebsiteDataStore.default().fetchDataRecords(ofTypes: [WKWebsiteDataTypeCookies]) { (_) in}
        var deferreds: [Deferred<Bool>] = []
        cookies.forEach({
            let deferred = Deferred<Bool>()
            deferreds.append(deferred)
            cookieStore.setCookie($0, completionHandler: {deferred.fill(true)})
        })
        return all(deferreds)
    }` 

plus the code change required at consumer side.  This does not improve the readability neither the lines of code. 
It basically performance required vs style.